libndt7/src/main/java/io/ooni/mk/nuvolari/libndt7/Client.java

```diff
@@ -1,100 +1,155 @@
+    private long t0 = 0;
+    private long tLast = 0;
+    private final long measurementInterval = TimeUnit.NANOSECONDS.convert(250, TimeUnit.MILLISECONDS);
+
+    @NonNull
+    private static final String TAG = "Client";
+
+    public Client(@NonNull Settings settings) {
         this.settings = settings;
     }
 
-    public void onLogInfo(String message) {
+    public void onLogInfo(@Nullable String message) {
+        Log.d(TAG, "onLogInfo: " + message);
         // NOTHING
     }
 
-    public void onError(String error) {
+    public void onError(@Nullable String error) {
+        Log.d(TAG, "onError: " + error);
         // NOTHING
     }
 
-    public void onServerDownloadMeasurement(Measurement measurement) {
+    public void onServerDownloadMeasurement(@NonNull Measurement measurement) {
+        Log.d(TAG, "onServerDownloadMeasurement: " + measurement);
         // NOTHING
     }
 
-    public void onClientDownloadMeasurement(Measurement measurement){
+    public void onClientDownloadMeasurement(@NonNull Measurement measurement){
+        Log.d(TAG, "onClientDownloadMeasurement: " + measurement);
         // NOTHING
     }
 
     @Override
     public final void onOpen(WebSocket ws, Response resp) {
-        onLogInfo("WebSocket connection established");
+        onLogInfo("WebSocket onOpen");
     }
 
     @Override
-    public final void onMessage(WebSocket ws, String text) {
+    public final void onMessage(WebSocket ws,
+                                String text) {
+        onLogInfo("WebSocket onMessage");
         count += text.length();
         periodic();
 
         Measurement measurement = new Measurement();
-        JSONObject doc;
+
+        @NonNull
+        final JSONObject doc;
+
+        // Root
         try {
             doc = new JSONObject(text);
-            measurement.elapsed = doc.getDouble("elapsed");
-            measurement.numBytes = doc.getDouble("num_bytes");
-        } catch (JSONException exc) {
-            // TODO(bassosimone): how to handle this failure? This is a case
-            // where the JSON has unexpected fields.
+        } catch (JSONException e) {
+            Log.d(TAG, "could not parse root json", e);
             return;
         }
+
+        // Top level
         try {
-            JSONObject bbrInfo = doc.getJSONObject("bbr_info");
-            measurement.bbrInfo = new BBRInfo();
-            try {
-                measurement.bbrInfo.bandwidth = bbrInfo.getDouble("bandwidth");
-                measurement.bbrInfo.RTT = bbrInfo.getDouble("rtt");
-            } catch (JSONException exc) {
-                // TODO(bassosimone): how to handle this failure? Like above
-                // this is a case where the JSON has unexpected fields.
-                return;
-            }
-        } catch (JSONException exc) {
-            // NOTHING: bbrInfo is optional
+            measurement.elapsed = doc.getDouble("elapsed");
+        } catch (JSONException e) { }
```

Comment:
I think it's an error if this field is missing

Reply:
Gotcha.

I think we should eventually do the JSON parsing differently, because this built-in deserialization is slow and awkward.  We should use Gson or another well-used library and/or require/allow the consumer to inject their own deserializer.

Reply:
I think Gson is fine